proposals/fixed-sized-buffers.md

```diff
@@ -29,42 +31,51 @@ public fixed DXGI_RGB GammaCurve[1025];
-struct <GammaCurve>e__FixedBuffer
+
+[FixedBuffer(typeof(DXGI_RGB), 1024)]
+public ConsoleApp1.<Buffer>e__FixedBuffer_1024<DXGI_RGB> GammaCurve;
+
+// Pack = 0 is the default packing and should result in indexable layout.
+[CompilerGenerated, UnsafeValueType, StructLayout(LayoutKind.Sequential, Pack = 0)]
+struct <Buffer>e__FixedBuffer_1024<T>
 {
-    private DXGI_RGB _e0;
-    private DXGI_RGB _e1;
+    private T _e0;
+    private T _e1;
     // _e2 ... _e1023
-    private DXGI_RGB _e1024;
+    private T _e1024;
 
-    public ref DXGI_RGB this[int index]
-    {
-        get;
-    }
+    public ref T this[int index] => ref (uint)index <= 1024u ?
+                                         ref RefAdd<T>(ref _e0, index):
+                                         throw new IndexOutOfRange();
 }
 ```
 
+Since such fixed-sized buffers no longer require use of `fixed`, it makes sense to allow any element type.  
+
+> NOTE: `fixed` will still be supported, but only if the element type is `blittable`
+
 ## Drawbacks
 [drawbacks]: #drawbacks
 
-Some additional syntax may be needed to ensure we don't break users that are currently using 'unsafe fixed-buffers'.
-* Given that the existing fixed-sized buffers only work with a selection of primitive types, it should be possible for the compiler to continue "just-working" if the user treats the fixed-buffer as a pointer.
-* Its probably worth extending the ability to work with fixed-sized buffers as pointers to all `blittable` types.
-
-In order to ensure that the safe fixed-sized buffer is resilient to future field-layout changes (and for it to work with non-blittable types), the compiler needs to explicitly layout each field of the array.
-* This can quickly become 'unwieldly' for large arrays
-* The JIT may not behave well with types that contain a large number of fields
- * We may be able to work with the JIT/Runtime to provide a special flag they can use (basically, it would give the contained type and the element count, then they just need to compute the size of the base struct and multiply, rather than actually checking each field).
+* There could be some challenges with backwards compatibility, but given that the existing fixed-sized buffers only work with a selection of primitive types, it should be possible for the compiler to continue "just-working" if the user treats the fixed-buffer as a pointer. 
+* Incompatible constructs may need to use slightly different `v2` encoding to hide the fields from old compiler.
+* Packing is not weel defined in IL spec for generic types. While the approach should work, we will be bordering on undocumnted behavior. We should make that documented and make sure other JITs like Mono have the same behavior.
+* Specifying a separate type for every length (an possibly another for `readonly` fields, if supported) will have impact on metadata. It will be bound by the number of arrays of different sizes in the given app. 
```

Comment:
It might be possible to limit this somewhat by declaring well known sizes and constructing buffers of other sizes from the small ones.  Ex: `struct <Buffer>e__FixedBuffer_1024<T>` could contain four elements of `struct <Buffer>e__FixedBuffer_256<T>`, rather than 1024 elements of `T`.